### PR TITLE
bug fixes plus some tests

### PR DESCRIPTION
**Pull request by omeshp**

### CODE REVIEW

File: Extensions/ArtifactEngine/Providers/filesystemProvider.ts [hunk 1]
```diff
@@ -33,21 +33,26 @@ export class FilesystemProvider implements models.IArtifactProvider {
...
             this.ensureDirectoryExistence(folder);
 
-            Logger.logMessage('Downloading '+ item.path + ' to '+ outputFilename);
+            Logger.logMessage('Downloading ' + item.path + ' to ' + outputFilename);
             const outputStream = fs.createWriteStream(outputFilename);
             stream.pipe(outputStream);
             stream.on("end",
                 () => {
                     Logger.logMessage(`Downloaded '${item.path}' to '${outputFilename}'`);
+                    if (!item.metadata) {
```
Comment (GitHubSriramB) [inline Extensions/ArtifactEngine/Providers/filesystemProvider.ts]: when will this happen?

File: Extensions/ArtifactEngine/Providers/webProvider.ts [hunk 2]
```diff
@@ -37,12 +38,27 @@ export class WebProvider implements models.IArtifactProvider {
...
+        var promise = new Promise<NodeJS.ReadableStream>(async (resolve, reject) => {
+            if (!artifactItem.metadata || !artifactItem.metadata['downloadUrl']) {
+                reject("No downloadUrl available to download the item.");
+            }
+
             var itemUrl: string = artifactItem.metadata['downloadUrl'];
             itemUrl = itemUrl.replace(/([^:]\/)\/+/g, "$1");
-            let res: httpm.HttpClientResponse = await this.httpc.get(itemUrl);
-            resolve(res.message);
+            var promise = this.httpc.get(itemUrl);
```
Comment (GitHubSriramB) [inline Extensions/ArtifactEngine/Providers/webProvider.ts]: don't we need to add any charset utf8 headers in the get call?

File: Extensions/ArtifactEngine/Engine/artifactEngine.ts [hunk 3]
```diff
@@ -96,4 +96,9 @@ export class ArtifactEngine {
 
     private artifactItemStore: ArtifactItemStore = new ArtifactItemStore();
     private logger: Logger;
-}
\ No newline at end of file
+}
+
+process.on('unhandledRejection', (reason, promise) => {
```
Comment (GitHubSriramB) [inline Extensions/ArtifactEngine/Engine/artifactEngine.ts]: what is this fix for?

File: Extensions/ArtifactEngine/EngineTests/artifactEngineTests.ts [hunk 4]
```diff
@@ -10,7 +11,7 @@ mockery.enable({
 mockery.registerMock('fs', {
     createWriteStream: (a) => {
         var mockedStream = stream.Writable();
-        mockedStream._write = () => {};
+        mockedStream._write = () => { };
```
Comment (GitHubSriramB) [inline Extensions/ArtifactEngine/EngineTests/artifactEngineTests.ts]: does the additional space matter?

File: Extensions/ArtifactEngine/Providers/filesystemProvider.ts [hunk 5]
```diff
@@ -19,8 +19,8 @@ export class FilesystemProvider implements models.IArtifactProvider {
         return this.getItems(itemsPath, artifactItem.path);
     }
 
-    getArtifactItem(artifactItem: models.ArtifactItem): Promise<Stream.Readable> {
-        var promise = new Promise<Stream.Readable>(async (resolve, reject) => {
+    getArtifactItem(artifactItem: models.ArtifactItem): Promise<NodeJS.ReadableStream> {
```
Comment (GitHubSriramB) [inline Extensions/ArtifactEngine/Providers/filesystemProvider.ts]: what is this change for?

File: Extensions/ArtifactEngine/Store/artifactItemStore.ts [hunk 6]
```diff
@@ -30,7 +30,7 @@ export class ArtifactItemStore {
 
     public itemsPendingProcessing(): boolean {
         var processingTickets = this._downloadTickets.filter(x => x.state === models.TicketState.Processing || x.state === models.TicketState.InQueue);
-        return processingTickets ? processingTickets.length != 0 : false;
+        return processingTickets.length != 0;
```
Comment (GitHubSriramB) [inline Extensions/ArtifactEngine/Store/artifactItemStore.ts]: wasn't the previous code safer?

File: Extensions/ArtifactEngine/task.ts [hunk 7]
```diff
@@ -75,6 +77,25 @@ async function downloadVSTSDropWithMultipleFiles(processorOptions) {
     await processor.processItems(webProvider, localFileProvider, processorOptions);
 }
 
+async function downloadVSTSDropWithMultipleFiles2(processorOptions) {
```
Comment (GitHubSriramB) [inline Extensions/ArtifactEngine/task.ts]: what is the difference in this?

File: Extensions/ArtifactEngine/package.json [hunk 8]
```diff
@@ -9,19 +9,21 @@
   },
```
Comment (GitHubSriramB) [inline Extensions/ArtifactEngine/package.json]: looks like i need to again register for the components we are using now :( are you sure these changes are required?

File: Extensions/ArtifactEngine/EngineTests/artifactEngineTests.ts [hunk 9]
```diff
@@ -51,26 +52,18 @@ describe('artifactEngine.processItems', () => {
     it('should call getArtifactItem only for artifact that match the download pattern', async () => {
         var testProvider = new providers.StubProvider();
         var downloadOptions = new engine.ArtifactEngineOptions();
-        downloadOptions.itemPattern = '*(PAth4|path5)\\**';
+        downloadOptions.itemPattern = '@(PAth4|path5)\\**';
```
Comment (GitHubSriramB) [inline Extensions/ArtifactEngine/EngineTests/artifactEngineTests.ts]: is this to validate the scenario of service passing multiple items to download for partial artifacts (the discussion we had with Lova)?

**Review by GitHubSriramB — APPROVED** [review]

File: Extensions/ArtifactEngine/Providers/webProvider.ts [hunk 10]
```diff
@@ -37,12 +38,27 @@ export class WebProvider implements models.IArtifactProvider {
...
+            promise.catch(reason => {
+                reject(reason);
+            });
+
+            let res: httpm.HttpClientResponse = await promise;
+
+            if (res.message.headers['content-encoding'] === 'gzip') {
```
Comment (GitHubSriramB) [inline Extensions/ArtifactEngine/Providers/webProvider.ts]: for build artifact, were we always getting gzip or is it conditional? wondering how we didn't find this issue so far?

File: Extensions/ArtifactEngine/Engine/artifactEngine.ts (hunk 3 above)
Comment (omeshp) [reply Extensions/ArtifactEngine/Engine/artifactEngine.ts]: if the code fails at some point in node like de-referencing undefined object it generates an unhandledRejection warning, we need to map it to error to fail in such cases.

File: Extensions/ArtifactEngine/EngineTests/artifactEngineTests.ts (hunk 4 above)
Comment (omeshp) [reply Extensions/ArtifactEngine/EngineTests/artifactEngineTests.ts]: no, but when i autoformat in vscode it generates this space.

File: Extensions/ArtifactEngine/EngineTests/artifactEngineTests.ts (hunk 9 above)
Comment (omeshp) [reply Extensions/ArtifactEngine/EngineTests/artifactEngineTests.ts]: No, it is just to test single patterns * will match Path4Path4 also.

File: Extensions/ArtifactEngine/Providers/filesystemProvider.ts (hunk 5 above)
Comment (omeshp) [reply Extensions/ArtifactEngine/Providers/filesystemProvider.ts]: After applying unzip transformation what we get is a ReadableStream which is the base call for Stream.Readable, we need to change the signatures to accommodate for that.

File: Extensions/ArtifactEngine/Providers/filesystemProvider.ts (hunk 1 above)
Comment (omeshp) [reply Extensions/ArtifactEngine/Providers/filesystemProvider.ts]: depened on the template currently all templates have metadata, if there is no metadata we should explicitly reject here.

File: Extensions/ArtifactEngine/Providers/webProvider.ts (hunk 2 above)
Comment (omeshp) [reply Extensions/ArtifactEngine/Providers/webProvider.ts]: not required, for the testing I did. Will take another look to check if it is required or not later.

File: Extensions/ArtifactEngine/Providers/webProvider.ts (hunk 10 above)
Comment (omeshp) [reply Extensions/ArtifactEngine/Providers/webProvider.ts]: It is conditional, I think it depends on size and api version but need to confirm.

File: Extensions/ArtifactEngine/Store/artifactItemStore.ts (hunk 6 above)
Comment (omeshp) [reply Extensions/ArtifactEngine/Store/artifactItemStore.ts]: The code was dead code, filter never returns undefined. It was caught using codecoverage.

**Review by GitHubSriramB — APPROVED** [review]

**Pull request merged**